test/jdk/java/util/Timer/AutoStop.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @requires !vm.gc.Epsilon
+ * @summary Discarding a Timer causes the Timer thread to stop.
+ */
+
+import java.util.Timer;
+import java.util.TimerTask;
+import java.lang.ref.Reference;
+
+public class AutoStop {
+    static final Object wakeup = new Object();
+    static volatile Thread tdThread = null;
+    static volatile int counter = 0;
+    static final int COUNTER_LIMIT = 10;
+
+    public static void main(String[] args) throws Exception {
+        Timer t = new Timer();
+
+        // Run an event that records the timer thread.
+        t.schedule(new TimerTask() {
+                public void run() {
+                    tdThread = Thread.currentThread();
+                    synchronized(wakeup) {
+                        wakeup.notify();
+                    }
+                }
+            }, 0);
+
+        // Wait for the thread to be accessible.
+        try {
+            synchronized(wakeup) {
+                while (tdThread == null) {
+                    wakeup.wait();
+                }
+            }
+        } catch (InterruptedException e) {
+        }
+
+        // Schedule some events that increment the counter.
+        for (int i = 0; i < COUNTER_LIMIT; ++i) {
+            t.schedule(new TimerTask() {
+                    public void run() {
+                        ++counter;
```

Comment:
This is not thread-safe. Operations on volatile variables are not atomic.

Reply:
Only one thread (the timer's thread) is writing, via a sequential series of task executions, so the simple increments are fine.  I made `counter` volatile because it's being written by one thread and read by a different thread.  Happy to drop the qualifier if that's not needed.

Reply:
Apologies - I thought there was real concurrency going on there. :)
The volatile is correct for the reason you cited.